### DISCUSSION TITLE
Allow / re-use LDAP filter syntax on included plug-ins in features

### DISCUSSION THREAD

**Discussion by thahnen** [eclipse-pde/eclipse.pde#1026] [Ideas]
## Summary

Currently, when creating bundles for Eclipse plug-ins/fragments there is the possibility to set a platform filter via the `Eclipse-PlatformFilter` header in the manifest file using the LDAP filter syntax which allows for complex conditions.
There is no such option for features when selecting plug-ins/fragments and even further features to be included.

## Why

We noticed, even though this is a corner case, that when creating a plug-in with fragments we can cover most of the environment combinations via specific fragments, we want to have a "fallback" fragment for every other environment combination. Setting it on the manifest file in the bundle itself works great having the freedom of filtering this way but when trying to put everything into a feature, the limitations are clearly shown.

## Idea

Having the LDAP filter syntax on features as well for included plug-ins/fragments/features would give us more freedom when creating the feature without the need to create multiple features (which would be possible but is way more overhead).

Additionally, it would be great that when including a fragment into one feature there would be the possibility to re-use the `Eclipse-PlatformFilter` header to configure the details regarding OS/window system/language/architecture.

**Comment by laeubi** [comment]
The usual pattern for this is to have everything in the feature but mark it as optional, that way items where filter do not match get excluded, would that work for you?

**Reply by thahnen** [reply]
Marking the "fallback" fragment as optional? Is that even possible?

**Comment by thahnen** [comment]
To make it clearer: It also applies if I have a bundle with a complex `Eclipse-PlatformFilter` configuration, that cannot be mapped into the feature configuration.

> Eclipse-PlatformFilter: (| (& (osgi.os=win32) (osgi.arch=x86)) (& (osgi.os=linux) (osgi.arch=x86_64)) (& (osgi.os=macosx) (osgi.arch=aarch64)))

Of course, this is no real-world example but I hope it shows the point.

**Comment by merks** [comment]
This is how fragments are included in features:
```
   <plugin
         id="org.eclipse.swt"
         version="0.0.0"/>

   <plugin
         id="org.eclipse.swt.win32.win32.x86_64"
         os="win32"
         ws="win32"
         arch="x86_64"
         version="0.0.0"/>
```
Like this in the structured editor:

![image](https://github.com/eclipse-pde/eclipse.pde/assets/208716/fcad07f3-ef5a-4480-876a-306ab04fb5f9)

You need something even more generalized than that?

**Reply by thahnen** [reply]
I know how it is done but it is confusing that it is different on the `Eclipse-PlatformFilter` (the actual fragment) and on the feature. I wasn't aware of the `p2.inf` file until you mentioned it, so maybe it is possible this way.

**Comment by merks** [comment]
Note that you can do even more complex things using a p2.inf via ` requires.<#>.filter = <ldap filter> (optional)`:

https://wiki.eclipse.org/Equinox/p2/Customizing_Metadata

If you do that for a feature, be sure to specify it in the build.properties:

![image](https://github.com/eclipse-pde/eclipse.pde/assets/208716/d64d02bd-3196-4aab-82c0-459a666a7cf6)
